pkg/server/plugin/keymanager/gcpkms/gcpkms.go

```diff
@@ -0,0 +1,1040 @@
+}
+
+// SignData creates a digital signature for the data to be signed.
+func (p *Plugin) SignData(ctx context.Context, req *keymanagerv1.SignDataRequest) (*keymanagerv1.SignDataResponse, error) {
+	if req.KeyId == "" {
+		return nil, status.Error(codes.InvalidArgument, "key id is required")
+	}
+	if req.SignerOpts == nil {
+		return nil, status.Error(codes.InvalidArgument, "signer opts is required")
+	}
+
+	p.entriesMtx.RLock()
+	defer p.entriesMtx.RUnlock()
+
+	keyEntry, hasKey := p.entries[req.KeyId]
+	if !hasKey {
+		return nil, status.Errorf(codes.NotFound, "key %q not found", req.KeyId)
+	}
+
+	var (
+		hashAlgo keymanagerv1.HashAlgorithm
+		digest   *kmspb.Digest
+	)
+	switch opts := req.SignerOpts.(type) {
+	case *keymanagerv1.SignDataRequest_HashAlgorithm:
+		hashAlgo = opts.HashAlgorithm
+	case *keymanagerv1.SignDataRequest_PssOptions:
+		// RSASSA-PSS is not supported by this plugin.
+		// See the comment in cryptoKeyVersionAlgorithmFromKeyType function for
+		// more details.
+		return nil, status.Error(codes.InvalidArgument, "the only RSA signature scheme supported is RSASSA-PKCS1-v1_5")
+	default:
+		return nil, status.Errorf(codes.InvalidArgument, "unsupported signer opts type %T", opts)
+	}
+	switch {
+	case hashAlgo == keymanagerv1.HashAlgorithm_UNSPECIFIED_HASH_ALGORITHM:
+		return nil, status.Error(codes.InvalidArgument, "hash algorithm is required")
+	case hashAlgo == keymanagerv1.HashAlgorithm_SHA256:
+		digest = &kmspb.Digest{
+			Digest: &kmspb.Digest_Sha256{Sha256: req.Data},
+		}
+	case hashAlgo == keymanagerv1.HashAlgorithm_SHA384:
+		digest = &kmspb.Digest{
+			Digest: &kmspb.Digest_Sha384{Sha384: req.Data},
+		}
+	default:
+		return nil, status.Error(codes.InvalidArgument, "hash algorithm not supported")
+	}
+
+	signResp, err := p.kmsClient.AsymmetricSign(ctx, &kmspb.AsymmetricSignRequest{
+		Name:   keyEntry.cryptoKeyVersionName,
+		Digest: digest,
+	})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to sign: %v", err)
+	}
+
+	// Perform integrity verification.
+	if int64(crc32Checksum(signResp.Signature)) != signResp.SignatureCrc32C.Value {
+		return nil, status.Error(codes.Internal, "error signing: response corrupted in-transit")
+	}
+
+	return &keymanagerv1.SignDataResponse{
+		Signature:      signResp.Signature,
+		KeyFingerprint: keyEntry.publicKey.Fingerprint,
+	}, nil
+}
+
+// createKey creates a new CryptoKey with a new CryptoKeyVersion.
+// If there is a specified IAM policy through the KeyPolicyFile configuration,
+// that policy is set to the created CryptoKey. If there is no IAM policy specified,
+// a default policy is constructed and attached. This function requests Cloud KMS
+// to get the public key of the created CryptoKeyVersion. A keyEntry is returned
+// with the CryptoKey, CryptoKeyVersion and public key.
+func (p *Plugin) createKey(ctx context.Context, spireKeyID string, keyType keymanagerv1.KeyType) (*keyEntry, error) {
+	p.entriesMtx.Lock()
+	defer p.entriesMtx.Unlock()
```

Comment:
Assuming I understand this correctly ... we'll create a key here when SPIRE prepares a new one ... but, we take a write lock here to do it, so the act of preparing a new key will block signing operations on our other (currently active) key?

Reply:
That's a good point. The lock window could be reduced using a function that would update the cache holding the write lock only in that function. This function could then call it when it updates the key entries cache. I'll make that change.